HPCC-20458 Allow File Access keys to be defined in the environment

kenrowland opened the pull request:
Add new CLI envmod for processing modification templates to modify
environment files.
Update modification template support.
Add unit tests for new features.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

Refactoring and updates

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b598298c14426e58d87dec5401453854aba152bf
Number of Error(s): 8
/mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp:57:64: error: invalid conversion from \xe2\x80\x98char\xe2\x80\x99 to \xe2\x80\x98const char*\xe2\x80\x99 [-fpermissive]

[Build error:](https://gist.github.com/e913dd5ae5f3dc7ac752ea89c2dcc90d)
Build log [Build log](https://gist.github.com/HPCCSmoketest/e913dd5ae5f3dc7ac752ea89c2dcc90d#file-result-19-01-07-15-37-26-log)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b598298c14426e58d87dec5401453854aba152bf
Number of Error(s): 8
/mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/XMLEnvironmentMgr.cpp:57:64: error: invalid conversion from \xe2\x80\x98char\xe2\x80\x99 to \xe2\x80\x98const char*\xe2\x80\x99 [-fpermissive]

[Build error:](https://gist.github.com/40096a5fcf88bfed0a1412d5c331b876)
Build log [Build log](https://gist.github.com/HPCCSmoketest/40096a5fcf88bfed0a1412d5c331b876#file-result-19-01-08-06-14-51-log)

HPCCSmoketest commented:
This PR (12027) is in the 1/2 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 10f82b4a99bd7e623a0f121c9799aad3865623e6
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |95 |1 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...

Backtrace of ./core_unittests.20389
[New LWP 20389]
[Thread debugging using libthread_db enabled]
Using host libthread_db library /lib64/libthread_db.so.1.
Core was generated by `hpcc/opt/HPCCSystems/bin/unittests -e ConfigMgrTemplateTests.
Program terminated with signal 11, Segmentation fault.
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
672	      { return _M_pi != 0 ? _M_pi->_M_get_use_count() : 0; }

Thread 1 (Thread 0x7f3db0c37140 (LWP 20389)):
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
#1  expired (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:1277
#2  lock (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr.h:497
#3  EnvironmentNode::getRoot (this=this@entry=0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:496
#4  0x00007f3db0482080 in EnvironmentNode::doFetchNodes (this=this@entry=0x0, configPath=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:397
#5  0x00007f3db0482b42 in EnvironmentNode::fetchNodes (this=0x0, path=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:385
#6  0x00007f3db04420c9 in EnvironmentMgr::fetchNodes (this=this@entry=0x87b0a0, path=/Root/Child1, nodes=std::vector of length 0, capacity 0, pStartNode=std::shared_ptr (empty) 0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp:437
#7  0x00007f3db04bee63 in Operation::getParentNodeIds (this=this@entry=0x875b68, pEnvMgr=pEnvMgr@entry=0x87b0a0, pInputs=pInputs@entry=0x893198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:102
#8  0x00007f3db04bfbc0 in OperationCreateNode::doExecute (this=0x875b68, pEnvMgr=0x87b0a0, pInputs=0x893198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/OperationCreateNode.cpp:33
#9  0x00007f3db04be13f in Operation::execute (this=0x875b68, pEnvMgr=0x87b0a0, pInputs=pInputs@entry=0x893198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:61
#10 0x00007f3db04a0900 in EnvModTemplate::execute (this=0x893180) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:442
#11 0x000000000047387b in ConfigMgrTemplateTests::executeTemplate (this=0x7b8c10, expectFailure=<optimized out>) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:168
#12 0x0000000000473b74 in ConfigMgrTemplateTests::Test_OperationsTests (this=0x7b8c10) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:471
#13 0x00007f3db01dd0c2 in CppUnit::TestCaseMethodFunctor::operator()() const () from /lib64/libcppunit-
 ... ( comment is too long, 77 lines, 5735 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  188 sec (00:03:08) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1495 sec (00:24:55) |   19 sec (00:00:19) | 1738 sec (00:28:58) |

```

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5d17fb20e711b5b6fdff7cc377619b413685cd23
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |95 |1 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...

Backtrace of ./core_unittests.22494
[New LWP 22494]
[Thread debugging using libthread_db enabled]
Using host libthread_db library /lib64/libthread_db.so.1.
Core was generated by `hpcc/opt/HPCCSystems/bin/unittests -e ConfigMgrTemplateTests.
Program terminated with signal 11, Segmentation fault.
#0  EnvironmentNode::getRoot (this=this@entry=0x0) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:676
676	      { return std::less<_Sp_counted_base<_Lp>*>()(this->_M_pi, __rhs._M_pi); }

Thread 1 (Thread 0x7f50e9560140 (LWP 22494)):
#0  EnvironmentNode::getRoot (this=this@entry=0x0) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:676
#1  0x00007f50e8dab000 in EnvironmentNode::doFetchNodes (this=this@entry=0x0, configPath=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:397
#2  0x00007f50e8dabac2 in EnvironmentNode::fetchNodes (this=0x0, path=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:385
#3  0x00007f50e8d6b0c9 in EnvironmentMgr::fetchNodes (this=this@entry=0x19080a0, path=/Root/Child1, nodes=std::vector of length 0, capacity 0, pStartNode=std::shared_ptr (empty) 0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp:437
#4  0x00007f50e8de7de3 in Operation::getParentNodeIds (this=this@entry=0x1902b68, pEnvMgr=pEnvMgr@entry=0x19080a0, pInputs=pInputs@entry=0x1920198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:102
#5  0x00007f50e8de8b40 in OperationCreateNode::doExecute (this=0x1902b68, pEnvMgr=0x19080a0, pInputs=0x1920198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/OperationCreateNode.cpp:33
#6  0x00007f50e8de70bf in Operation::execute (this=0x1902b68, pEnvMgr=0x19080a0, pInputs=pInputs@entry=0x1920198) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:61
#7  0x00007f50e8dc9880 in EnvModTemplate::execute (this=0x1920180) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:442
#8  0x000000000047387b in ConfigMgrTemplateTests::executeTemplate (this=0x1845c10, expectFailure=<optimized out>) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:168
#9  0x0000000000473b74 in ConfigMgrTemplateTests::Test_OperationsTests (this=0x1845c10) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:471
#10 0x00007f50e8b060c2 in CppUnit::TestCaseMethodFunctor::operator()() const () from /lib64/libcppunit-1.12.so.1
#11 0x00007f50e8afc293 in CppUnit::DefaultProtector::protect(CppUnit::Functor const&, CppUnit::ProtectorContext const&) () from /lib64/libcppunit-1.12.so.1
#12 0x00007f50e8b0333d in CppUnit::ProtectorChain::protect(CppUnit::Functor const&, Cp
 ... ( comment is too long, 74 lines, 5481 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   40 sec (00:00:40) |  184 sec (00:03:04) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1469 sec (00:24:29) |   20 sec (00:00:20) | 1733 sec (00:28:53) |

```

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 59eaebc807714e5d62101a23910ee9cfa720d42a
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |95 |1 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...

Backtrace of ./core_unittests.23505
[New LWP 23505]
[Thread debugging using libthread_db enabled]
Using host libthread_db library /lib64/libthread_db.so.1.
Core was generated by `hpcc/opt/HPCCSystems/bin/unittests -e ConfigMgrTemplateTests.
Program terminated with signal 11, Segmentation fault.
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
672	      { return _M_pi != 0 ? _M_pi->_M_get_use_count() : 0; }

Thread 1 (Thread 0x7f4e0110f140 (LWP 23505)):
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
#1  expired (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:1277
#2  EnvironmentNode::getRoot (this=this@entry=0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:496
#3  0x00007f4e0095a000 in EnvironmentNode::doFetchNodes (this=this@entry=0x0, configPath=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:397
#4  0x00007f4e0095aac2 in EnvironmentNode::fetchNodes (this=0x0, path=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:385
#5  0x00007f4e0091a0c9 in EnvironmentMgr::fetchNodes (this=this@entry=0x9a0d00, path=/Root/Child1, nodes=std::vector of length 0, capacity 0, pStartNode=std::shared_ptr (empty) 0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp:437
#6  0x00007f4e00996de3 in Operation::getParentNodeIds (this=this@entry=0x9c1808, pEnvMgr=pEnvMgr@entry=0x9a0d00, pInputs=pInputs@entry=0x9acb28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:102
#7  0x00007f4e00997b40 in OperationCreateNode::doExecute (this=0x9c1808, pEnvMgr=0x9a0d00, pInputs=0x9acb28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/OperationCreateNode.cpp:33
#8  0x00007f4e009960bf in Operation::execute (this=0x9c1808, pEnvMgr=0x9a0d00, pInputs=pInputs@entry=0x9acb28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:61
#9  0x00007f4e00978880 in EnvModTemplate::execute (this=0x9acb10) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:442
#10 0x0000000000471fbc in executeTemplate (expectFailure=false, this=0x8ed780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:168
#11 ConfigMgrTemplateTests::Test_OperationsTests (this=0x8ed780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:471
#12 0x00007f4e006b50c2 in CppUnit::TestCaseMethodFunctor::operator()() const () from /lib64/libcppunit-1.12.so.1
#13 0x00007f4e006ab293 in CppUnit::DefaultProtector::protect(CppUnit::Functor const&, CppUnit::ProtectorContext con
 ... ( comment is too long, 76 lines, 5609 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1466 sec (00:24:26) |   19 sec (00:00:19) | 1720 sec (00:28:40) |

```

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3919127e6d5c06ac7df6e705e7fc846feff88b10
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |95 |1 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...

Backtrace of ./core_unittests.21395
[New LWP 21395]
[Thread debugging using libthread_db enabled]
Using host libthread_db library /lib64/libthread_db.so.1.
Core was generated by `hpcc/opt/HPCCSystems/bin/unittests -e ConfigMgrTemplateTests.
Program terminated with signal 11, Segmentation fault.
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
672	      { return _M_pi != 0 ? _M_pi->_M_get_use_count() : 0; }

Thread 1 (Thread 0x7ff3f101f140 (LWP 21395)):
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
#1  expired (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:1277
#2  lock (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr.h:497
#3  EnvironmentNode::getRoot (this=this@entry=0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:506
#4  0x00007ff3f08696d0 in EnvironmentNode::doFetchNodes (this=this@entry=0x0, configPath=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:407
#5  0x00007ff3f086a192 in EnvironmentNode::fetchNodes (this=0x0, path=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:395
#6  0x00007ff3f0829219 in EnvironmentMgr::fetchNodes (this=this@entry=0xa1bd40, path=/Root/Child1, nodes=std::vector of length 0, capacity 0, pStartNode=std::shared_ptr (empty) 0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp:437
#7  0x00007ff3f08a64b3 in Operation::getParentNodeIds (this=this@entry=0xa3c848, pEnvMgr=pEnvMgr@entry=0xa1bd40, pInputs=pInputs@entry=0xa27b68) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:102
#8  0x00007ff3f08a7210 in OperationCreateNode::doExecute (this=0xa3c848, pEnvMgr=0xa1bd40, pInputs=0xa27b68) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/OperationCreateNode.cpp:33
#9  0x00007ff3f08a578f in Operation::execute (this=0xa3c848, pEnvMgr=0xa1bd40, pInputs=pInputs@entry=0xa27b68) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:61
#10 0x00007ff3f0887f50 in EnvModTemplate::execute (this=0xa27b50) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:442
#11 0x0000000000471fbc in executeTemplate (expectFailure=false, this=0x968780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:168
#12 ConfigMgrTemplateTests::Test_OperationsTests (this=0x968780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:471
#13 0x00007ff3f05c40c2 in CppUnit::TestCaseMethodFunctor::operator()() const () from /lib64/libcppunit-1.12.so.1
#14 0x00007ff3f05ba293 in CppUnit::DefaultPro
 ... ( comment is too long, 77 lines, 5679 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   28 sec (00:00:28) |  187 sec (00:03:07) |    0 sec (00:00:00) |    2 sec (00:00:02) |   17 sec (00:00:17) | 1514 sec (00:25:14) |   19 sec (00:00:19) | 1767 sec (00:29:27) |

```

AttilaVamos commented:
@kenrowland Maybe I'm wrong, but this line looks like very suspicious to me:
```
#3  EnvironmentNode::getRoot (this=this@entry=0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:506
```
especially this part:  "this=this@entry=0x0"
As well as this:
```
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
```
where "this=0x30"

kenrowland commented:
@AttilaVamos one important thing to note is that this all works on Ubuntu

AttilaVamos commented:
@kenrowland I know.

AttilaVamos commented:
@kenrowland I should check something. Give me ~ half an hour.

kenrowland commented:
@AttilaVamos the failure on my last commit is a logic error. I did not test on Ubuntu, so consider it an invalid workaround.

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4f832099b5423d0513c2dfa96f4d2a7fccd165d2
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |95 |1 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...

Backtrace of ./core_unittests.10894
[New LWP 10894]
[Thread debugging using libthread_db enabled]
Using host libthread_db library /lib64/libthread_db.so.1.
Core was generated by `hpcc/opt/HPCCSystems/bin/unittests -e ConfigMgrTemplateTests.
Program terminated with signal 11, Segmentation fault.
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
672	      { return _M_pi != 0 ? _M_pi->_M_get_use_count() : 0; }

Thread 1 (Thread 0x7f161f967140 (LWP 10894)):
#0  _M_get_use_count (this=0x30) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:672
#1  expired (this=0x28) at /usr/include/c++/4.8.2/bits/shared_ptr_base.h:1277
#2  EnvironmentNode::getRoot (this=this@entry=0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:515
#3  0x00007f161f1b2000 in EnvironmentNode::doFetchNodes (this=this@entry=0x0, configPath=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:407
#4  0x00007f161f1b2ac2 in EnvironmentNode::fetchNodes (this=0x0, path=..., nodes=std::vector of length 0, capacity 0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentNode.cpp:395
#5  0x00007f161f1720c9 in EnvironmentMgr::fetchNodes (this=this@entry=0xa56d00, path=/Root/Child1, nodes=std::vector of length 0, capacity 0, pStartNode=std::shared_ptr (empty) 0x0) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/EnvironmentMgr.cpp:437
#6  0x00007f161f1eede3 in Operation::getParentNodeIds (this=this@entry=0xa77808, pEnvMgr=pEnvMgr@entry=0xa56d00, pInputs=pInputs@entry=0xa62b28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:102
#7  0x00007f161f1efb40 in OperationCreateNode::doExecute (this=0xa77808, pEnvMgr=0xa56d00, pInputs=0xa62b28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/OperationCreateNode.cpp:33
#8  0x00007f161f1ee0bf in Operation::execute (this=0xa77808, pEnvMgr=0xa56d00, pInputs=pInputs@entry=0xa62b28) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:61
#9  0x00007f161f1d0880 in EnvModTemplate::execute (this=0xa62b10) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:442
#10 0x0000000000471fbc in executeTemplate (expectFailure=false, this=0x9a3780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:168
#11 ConfigMgrTemplateTests::Test_OperationsTests (this=0x9a3780) at /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp:471
#12 0x00007f161ef0d0c2 in CppUnit::TestCaseMethodFunctor::operator()() const () from /lib64/libcppunit-1.12.so.1
#13 0x00007f161ef03293 in CppUnit::DefaultProtector::protect(CppUnit::Functor const&, CppUnit::ProtectorContext con
 ... ( comment is too long, 76 lines, 5609 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  193 sec (00:03:13) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1490 sec (00:24:50) |   19 sec (00:00:19) | 1749 sec (00:29:09) |

```

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fd7c4724c8673d9cedcc357907013b0412b82a3b
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |886 |886 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   44 sec (00:00:44) |  189 sec (00:03:09) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1507 sec (00:25:07) |   19 sec (00:00:19) | 1780 sec (00:29:40) |

AttilaVamos commented:
Well Done! (That's why we really need Smoketest and OBT.)

kenrowland commented:
@AttilaVamos I agree! That's why once I figured out how to implement unit tests, I started adding them for all my code. I still have more I will be adding as time goes by so that when I make a change to core code, it's easy to test it.

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 22b2501a8faaefd0dada8dd999702f848e6ee479
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |98 |3 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Unittest error(s):
```
ConfigMgr 2.0 - Load and Parse - Verifying configuration XSDs are compliant...
*** ConfigMgr 2.0 - Templates - Loading Only ***
*** ConfigMgr 2.0 - Templates - Inputs ***
*** ConfigMgr 2.0 - Templates - Variable substitution tests ***
*** ConfigMgr 2.0 - Templates - Execute Templates ***
*** ConfigMgr 2.0 - Templates - Find Node Tests ***
*** ConfigMgr 2.0 - Templates - Create Node Tests ***
Run:  6   Failures: 3   Errors: 0
1) test: ConfigMgrTemplateTests::Test_OperationsTests (F) line: 167 /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp
- Expression: rc
- Loading... There was an error parsing the JSON, offset: 0, error = The document is empty.
2) test: ConfigMgrTemplateTests::Test_FindNodeTests (F) line: 167 /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp
- Expression: rc
- Loading... There was an error parsing the JSON, offset: 0, error = The document is empty.
3) test: ConfigMgrTemplateTests::Test_CreateNodeTests (F) line: 167 /mnt/disk1/home/vamosax/smoketest/PR-12027/HPCC-Platform/testing/unittests/configmgr/ConfigMgrTemplateTests.cpp
- Expression: rc
- Loading... There was an error parsing the JSON, offset: 0, error = The document is empty.

```

Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |886 |886 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  185 sec (00:03:05) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1502 sec (00:25:02) |   19 sec (00:00:19) | 1752 sec (00:29:12) |

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 50b6cf88183b839d2d7c852a79456a80f4aac09e
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |886 |886 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   50 sec (00:00:50) |  186 sec (00:03:06) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1468 sec (00:24:28) |   20 sec (00:00:20) | 1744 sec (00:29:04) |

HPCCSmoketest commented:
Process of PR-12027, label: HPCC-20458 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 18247a11d80dddc9755b38ce9dc098f64388f951
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |886 |886 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   35 sec (00:00:35) |  185 sec (00:03:05) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1628 sec (00:27:08) |   20 sec (00:00:20) | 1890 sec (00:31:30) |

jpmcmu commented on configuration/configmgr/configmgrlib/mod_template_support/Variable.cpp:
```diff
@@ -82,14 +81,14 @@ std::string Input::getValue(size_t idx) const
-    }
-    else
-    {
+//    if (!m_values.empty())
```
Any reason to keep the commented out code?

kenrowland replied:
Um, no. I thought I had removed all the commented code, guess I missed this one. I will remove.

jpmcmu reviewed (commented):
@kenrowland Looks good, a couple of comments

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/Operation.cpp:
```diff
@@ -71,3 +73,110 @@ void Operation::assignAttributeCookedValues(Inputs *pInputs)
+        if (!attr.saveValue.empty())
+        {
+            bool found=false, set=false;
+            std::shared_ptr<Variable> pInput = pInputs->getVariable(attr.saveValue);
```
No. Previously the attr.saveValue member would have been used to add an entry by that name to the Variables object. However, the getVariable method will throw an exception if the requested variable does not exist.

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:
```diff
@@ -220,124 +282,195 @@ void EnvModTemplate::parseOperation(const rapidjson::Value &operation)
+    // Get the count (optional, default is 1)
+    it = dataObj.FindMember("count");
+    if (it != dataObj.MemberEnd())
+        pOp->m_count = trim(it->value.GetString());
```
No because they could be substitution variables in the form {{varname}}. For that reason, they are kept in their raw for until used whereby a substitution and conversion to a number is done. At that time an exception, during execution of the template, can be thrown if there is a problem.

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:
```diff
@@ -148,59 +151,118 @@ void EnvModTemplate::parseTemplate()
-    std::string inputName(input.FindMember("name")->value.GetString());
-    std::string type(input.FindMember("type")->value.GetString());
-    pInput = inputValueFactory(type, inputName);
+    std::string varName(varValue.FindMember("name")->value.GetString());
```
Yes, they are required by the schema.

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:
```diff
@@ -220,124 +282,195 @@ void EnvModTemplate::parseOperation(const rapidjson::Value &operation)
+    // Get the starting index (optional, for windowing into a range of values)
+    it = dataObj.FindMember("start_index");
+    if (it != dataObj.MemberEnd())
+        pOp->m_startIndex = trim(it->value.GetString());
```
See comment for m_count above.

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:
```diff
@@ -220,124 +282,195 @@ void EnvModTemplate::parseOperation(const rapidjson::Value &operation)
+    rapidjson::Value::ConstMemberIterator saveInfoIt = dataObj.FindMember("save_nodeid");
+    if (saveInfoIt != dataObj.MemberEnd())
+    {
+        pOp->m_saveNodeIdName = saveInfoIt->value.GetObject().FindMember("save_name")->value.GetString();
```
Yes, it is required by the schema if the save object exists.

kenrowland commented on configuration/configmgr/configmgrlib/mod_template_support/EnvModTemplate.cpp:
```diff
@@ -220,124 +282,195 @@ void EnvModTemplate::parseOperation(const rapidjson::Value &operation)
+            saveInfoIt = attr.FindMember("save_value");
+            if (saveInfoIt != attr.MemberEnd())
+            {
+                newAttribute.saveValue = saveInfoIt->value.GetObject().FindMember("save_name")->value.GetString();
```
The same sub schema definition is use here as well.

Commit:
HPCC-20458 Allow File Access keys to be defined in the environment

Add new CLI envmod for processing modification templates to modify
environment files.
Update modification template support.
Add unit tests for new features.

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 87af07935b61ec0ddc71d16fcd2edecfc0ef3f32
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |101 |101 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |738 |738 |0 |
|test (roxie) |886 |886 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  189 sec (00:03:09) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1468 sec (00:24:28) |   18 sec (00:00:18) | 1726 sec (00:28:46) |

jpmcmu commented:
@kenrowland looks good

kenrowland commented:
@richardkchapman Please merge
